### PR TITLE
Issue #260: Bicikelj panel

### PR DESCRIPTION
**Pull request by lazarm**
It's not entirely finished, I still need to do pushing and displaying changes, but I hope I'll get some help here :)

### CODE REVIEW

File: piplmesh/panels/bicikelj/management/commands/updatebicikelj.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
```
Comment (mitar) [inline piplmesh/panels/bicikelj/management/commands/updatebicikelj.py:0]: This is not really needed. As we will run task very often (on minute basis) there is no need for command. For horoscope, which runs only once per day, it is necessary so you can test.

File: piplmesh/settings.py [hunk 2]
```diff
@@ -189,6 +190,8 @@
     'tastypie_mongoengine',
     'sekizai',
     'missing',
+    'django_mongodb_engine',
```
Comment (mitar) [inline piplmesh/settings.py]: What is this?

File: piplmesh/settings.py [hunk 3]
```diff
@@ -189,6 +190,8 @@
...
+    'djangotoolbox',
```
Comment (mitar) [inline piplmesh/settings.py]: Why is this needed?

File: piplmesh/settings.py [hunk 4]
```diff
@@ -178,6 +178,7 @@
     'piplmesh.utils',
     'piplmesh.panels',
     'piplmesh.panels.horoscope', # To load manage.py command
+    'piplmesh.panels.bicikelj',
```
Comment (mitar) [inline piplmesh/settings.py]: You should not need this.

File: piplmesh/settings.py [hunk 5]
```diff
@@ -219,6 +222,7 @@
 
 CHECK_ONLINE_USERS_INTERVAL = 10 # seconds
 CHECK_FOR_NEW_HOROSCOPE = 6 # am every day
+CHECK_BICIKELJ_STANDS_INFO = 60
```
Comment (mitar) [inline piplmesh/settings.py]: Rename to `POLL_BICIKELJ_INTERVAL`.

Comment (mitar) [inline piplmesh/settings.py]: Add units into a comment at the end.

File: piplmesh/panels/bicikelj/models.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import mongoengine
+
+NUMBER_OF_STATIONS = 32
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: O no no. You do not need to hardcode that. You have everything in `carto` URL.

File: piplmesh/panels/bicikelj/models.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+class BicikeljStation(mongoengine.Document):
+    name = mongoengine.StringField(required=True)
+    address = mongoengine.StringField(required=True)
+    number = mongoengine.IntField(required=True)
+    position = mongoengine.GeoPointField(required=True)
+    available = mongoengine.IntField()
+    free = mongoengine.IntField()
+    open = mongoengine.IntField(choices=(0,1))
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Ehm, boolean is not OK? :-)

**Comment by mitar** [comment]
BTW, #248 is not Bicikelj panel.

File: piplmesh/panels/bicikelj/models.py (hunk 7 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Move `open` after `position`.

File: piplmesh/panels/bicikelj/models.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import mongoengine
+
+NUMBER_OF_STATIONS = 32
+
+class BicikeljStation(mongoengine.Document):
+    name = mongoengine.StringField(required=True)
+    address = mongoengine.StringField(required=True)
+    number = mongoengine.IntField(required=True)
+    position = mongoengine.GeoPointField(required=True)
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Rename to `location`.

File: piplmesh/panels/bicikelj/models.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import mongoengine
+
+NUMBER_OF_STATIONS = 32
+
+class BicikeljStation(mongoengine.Document):
+    name = mongoengine.StringField(required=True)
+    address = mongoengine.StringField(required=True)
+    number = mongoengine.IntField(required=True)
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Rename to `id`.

Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Move to very beginning.

File: piplmesh/panels/bicikelj/models.py (hunk 7 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Store also total. It does not matter that data is not normalized. It is easier to search like that.

File: piplmesh/panels/bicikelj/models.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    position = mongoengine.GeoPointField(required=True)
+    available = mongoengine.IntField()
+    free = mongoengine.IntField()
+    open = mongoengine.IntField(choices=(0,1))
+    ticket = mongoengine.IntField(choices=(0,1))
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Ignore this one.

File: piplmesh/panels/bicikelj/models.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    timestamp = mongoengine.IntField()
```
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Use datetime field.

File: piplmesh/panels/bicikelj/panel.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+from __future__ import absolute_import
+from django.utils.translation import ugettext_lazy as _
```
Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py:3]: Empty line between each group of imports.

File: piplmesh/panels/bicikelj/panel.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+class BicikeljPanel(panels.BasePanel):
+    def get_context(self, context):
+
+        context.update({
+            'header': _("Bicikelj stations"),
+        })
+
+        try:
+            node = self.request.node
+        except :
```
Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: This is not possible. Furthermore, never catch ALL exceptions, only those you expect.

File: piplmesh/panels/bicikelj/panel.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
+from __future__ import absolute_import
+from django.utils.translation import ugettext_lazy as _
+from piplmesh import panels
+from . import stations
+
+class BicikeljPanel(panels.BasePanel):
+    def get_context(self, context):
+
```
Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: Unnecessary empty line.

File: piplmesh/panels/bicikelj/stations.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+BICIKELJ_STANDS_URL = 'http://www.bicikelj.si/service/carto'
+BICIKELJ_INFO_URL = 'http://www.bicikelj.si/service/stationdetails/ljubljana/'
+ZOOM_LEVEL = 15
+MAP_WIDTH = 450
+C = 0.703119412486786 # constant at 60 degrees
+
+def max_distance(zoom, width):
+    return C*width/math.pow(2,zoom)
+
+def get_nearest_stations(lat, lng):
```
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: `get_nearest_stations` is not really a good name. Nearest is really only the one (it would be really strange coincidence to have more than one nearest). You are not search really for the nearest, but all in some close proximity, range.

File: piplmesh/panels/bicikelj/panel.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        stands = stations.get_nearest_stations(node.latitude, node.longitude)
+
+        if not stands:
+            context.update({
+                'error_data': True,
+            })
+            return context
+
+        context.update({
+            'stations': stands,
```
Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: Why once `stations` and then `stands`?

Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: You have problems with conflict with imported module? :-)

Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: So this function could really be very simple:

```
context.update({
    'header': _("Bicikelj stations"),
    'stations': stations.get_stations(self.request.node.latitude, self.request.node.longitude),
})
```

Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: And then you can simply in the template check if there is any data in `stations`. Which is not really necessary an error, simply maybe there are no stations around.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+{% block content %}
+    {% addtoblock "css" %}<link href="{% static "piplmesh/panel/bicikelj/panel.css" %}" rel="stylesheet" type="text/css" media="screen" />{% endaddtoblock %}
+    <div id="bicikelj">
+        {% if error_node %}
+        <span class="bicikelj_error">{% trans "Unknown location." %}</span>
+        {% elif error_data %}
+        <span class="bicikelj_error">{% trans "No Bicikelj stands nearby" %}</span>
+        {% else %}
+           {% for st in stations %}
```
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Write variable names out, fully.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+                <div id="'st+str(st.number)'">
```
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: What is this???

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+                <dl>
+                    <dt>
+                        <span class="station_name">{{ st.name }}</span>
+                    </dt>
+                    <dd>
+                        <ul>
+                            <li><span class="station_availability">{% trans "Available bikes: " %}<span></span></span></li>
```
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: `<span></span></span>`???

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+                            <li><span class="station_free">{% trans "Free places: " %}{{ st.free }}</span></li>
+                            <li><span class="station_ticket">{% trans "Credit card: " %}{% if st.ticket == 1 %}{% trans "yes" %}{% else %}{% trans "no" %}{% endif %}</span></li>
```
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Is there any with ticket==1 in Ljubljana?

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+           {% for st in stations %}
+                <div id="'st+str(st.number)'">
+                <dl>
+                    <dt>
+                        <span class="station_name">{{ st.name }}</span>
+                    </dt>
+                    <dd>
+                        <ul>
+                            <li><span class="station_availability">{% trans "Available bikes: " %}<span></span></span></li>
+                            <li><span class="station_free">{% trans "Free places: " %}{{ st.free }}</span></li>
```
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Hm, wouldn't it be better to have string "23 free places"? Can you format string like that?

File: piplmesh/panels/bicikelj/tasks.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
+from celery import task
+from pushserver.utils import updates
+
+from piplmesh.frontend import views
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py:5]: Group imports properly.

File: piplmesh/panels/bicikelj/tasks.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+from . import models, stations
+
+@task.task
+def update_station_info():
+    station_info = stations.fetch_data()
+    for n in range(1,models.NUMBER_OF_STATIONS+1):
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: Why? Why you cannot simply iterate over anything returned by `fetch_data`?

File: piplmesh/panels/bicikelj/tasks.py [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+        if not models.BicikeljStation.objects(number=n, timestamp=station_info[n-1]['timestamp']):
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: There is a racing condition here. You should not check and then save. Use upsert.

File: piplmesh/panels/bicikelj/tasks.py [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+            models.BicikeljStation(
+                                           name=station_info[n-1]['name'],
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: Indentation.

And there are easier ways to pass this data if keys in the dict are the same as keys in the document model. Which they should be.

File: piplmesh/panels/bicikelj/tasks.py [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+                                           timestamp=station_info[n-1]['timestamp'],
+                                           available = station_info[n-1]['available'],
+                                           free = station_info[n-1]['free'],
+                                           open = station_info[n-1]['open'],
+                                           ticket = station_info[n-1]['ticket'],
+            ).save()
+            updates.send_update(
+                views.HOME_CHANNEL_ID,
+                # TODO
+                )
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: Indentation.

File: piplmesh/panels/bicikelj/tasks.py [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+
+
```
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: Unnecessary empty lines.

File: piplmesh/panels/bicikelj/stations.py [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+            break
+    stands = [models.BicikeljStation.objects(number=num).order_by('-timestamp').first() for num in ids]
+    return stands
+
+def fetch_data():
+    stands_data = urllib2.urlopen(BICIKELJ_STANDS_URL).read()
+    stands_xml_data = minidom.parseString(stands_data)
+    stands_nodes = stands_xml_data.getElementsByTagName('marker')
+    info_data = [urllib2.urlopen(BICIKELJ_INFO_URL+str(number)).read() for number in range(1,models.NUMBER_OF_STATIONS+1)]
+    info_xml_data = [minidom.parseString(info_data[i]) for i in range(models.NUMBER_OF_STATIONS)]
```
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: Do it without using `NUMBER_OF_STATIONS`.

File: piplmesh/panels/bicikelj/stations.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+        except:
+            break
+    stands = [models.BicikeljStation.objects(number=num).order_by('-timestamp').first() for num in ids]
+    return stands
+
+def fetch_data():
+    stands_data = urllib2.urlopen(BICIKELJ_STANDS_URL).read()
+    stands_xml_data = minidom.parseString(stands_data)
+    stands_nodes = stands_xml_data.getElementsByTagName('marker')
+    info_data = [urllib2.urlopen(BICIKELJ_INFO_URL+str(number)).read() for number in range(1,models.NUMBER_OF_STATIONS+1)]
```
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: Use more spaces, for example, around `,`.

File: piplmesh/panels/bicikelj/stations.py [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+    while 1:
+        try:
+            ids.append(models.BicikeljStation.objects(position__near= [lat, lng], position__within_box=[(lat-abs(maxd),lng-abs(maxd)),(lat+abs(maxd),lng+abs(maxd))], number__nin=ids).first().number)
+        except:
+            break
+    stands = [models.BicikeljStation.objects(number=num).order_by('-timestamp').first() for num in ids]
+    return stands
+
+def fetch_data():
+    stands_data = urllib2.urlopen(BICIKELJ_STANDS_URL).read()
```
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: Decide if you name things `stations` or `stands`.

File: piplmesh/panels/bicikelj/stations.py [hunk 31]
```diff
@@ -0,0 +1,45 @@
+import math, urllib2
+from xml.dom import minidom
+
+from . import models
+
+BICIKELJ_STANDS_URL = 'http://www.bicikelj.si/service/carto'
+BICIKELJ_INFO_URL = 'http://www.bicikelj.si/service/stationdetails/ljubljana/'
+ZOOM_LEVEL = 15
+MAP_WIDTH = 450
+C = 0.703119412486786 # constant at 60 degrees
```
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: I think I said that you should not do it like that.

File: piplmesh/panels/bicikelj/panel.py (hunk 13 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/panel.py]: Does node also have a value if a user comes to our page when not connected to one of the wireless nodes?

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 17 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: So is this a right way, displaying info about each node in a for loop?

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 18 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Nothing yet :-) I don't know anything about jquery, but I was reading about accessing html elements and their values. And I don't know any other way how to  access a particular element in a for loop (for example free spots of the third station) than adding id value for each div element in a loop (but then I also read having a number as id is possible only for HTML5, but is not recommended). So I'm still without any ideas.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 20 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: No, I'll remove it from a model.

File: piplmesh/panels/bicikelj/tasks.py (hunk 24 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/tasks.py]: I know. My idea was to find a particular station with the newest timestamp, and if it fails, it would save it and push changes to the server. So pushing happens only if timestamp is changed. If I use upsert, it will update if it finds the object (which is pointless since timestamp is old and so is the data about free spots etc) or save it. So I would have to know  which action happened, update or save. to know if pushing needs to be done. Is there a way to retrieve this info?

File: piplmesh/panels/bicikelj/stations.py (hunk 30 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/stations.py]: You don't mind any of them?

File: piplmesh/panels/bicikelj/stations.py (hunk 31 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/stations.py]: I know, I just wanted to check if formula si correct and only stations within a map are chosen. But still, I don't quite understand how to do it. Should I calculate NW and SE points based on node and zoom level inside a panel.js in a map panel, and then use:

> var bounds = new google.maps.LatLngBounds();
> bounds.extend(NW);
> bounds.extend(SE);

And how should I access these two points to query the stations with "within_box" ?

**Comment by mitar** [comment]
Please update from main repository.

File: piplmesh/panels/bicikelj/panel.py (hunk 13 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/panel.py]: Yes. You can assume that `request.node` is always populated with something. How it is done you do not need to worry. :-)

(Currently, we allow only access inside the network.)

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 17 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Sure. Just write `station` instead of `st`. Use indentation correctly. Order them in some meaningful way (maybe by distance from the node).

File: piplmesh/panels/bicikelj/stations.py (hunk 30 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: I don't know. What they are using in their API? Use the same.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 18 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: You could do something like that, yes. But this syntax is not correct for Django templates.

Or, you could use `.data()` function from jQuery to attach arbitrary data to DOM elements. Like id. And then search for that. (We are using that in #199.

File: piplmesh/panels/bicikelj/tasks.py (hunk 24 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/tasks.py]: OK. You are right. In this context it does not matter if we override the previous value, because we will override with same content. So yes, we can check if value exists and if not, we store it.

Please rename `number` to `id`. And don't have number of stations hard-coded.

File: piplmesh/panels/bicikelj/stations.py (hunk 31 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: You store width and height (bounds) hard-coded as global variables in Python. You use that for "within_box". And pass them to JavaScript. And use them to check if they are inside of bounds when map is displayed.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 17 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: They are already ordered in get_nearest_stations (yes, I'll rename that function).  __near=[lat,lng]  orderes all stations by distance from the [lat,lng] and within_box removes those who fall out of map.

File: piplmesh/panels/bicikelj/management/commands/updatebicikelj.py (hunk 1 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/management/commands/updatebicikelj.py:0]: TODO.

File: piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html (hunk 17 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/templates/panel/bicikelj/panel.html]: Great!

**Comment by mitar** [comment]
So, waiting for you ... :-)

File: piplmesh/panels/bicikelj/stations.py (hunk 31 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/stations.py]: I don't understand your last sentence. Where and what should I check?

File: piplmesh/panels/bicikelj/models.py (hunk 9 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/models.py]: How do I query with id? models.BicikeljStation.objects(id=station['id'],..)  keeps returning ValidationError?

Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Do you have `id` field? Currently you have it named `number`. You should rename it to `id`. The idea is that `number` name is badly named.

File: piplmesh/panels/bicikelj/stations.py (hunk 31 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/stations.py]: That values you have computed are still in the bounds of the displayed map.

File: piplmesh/panels/bicikelj/models.py (hunk 9 above)
Comment (lazarm) [inline piplmesh/panels/bicikelj/models.py]: I renamed it in my code and then checked if everything works fine, but it returns ValidationError, when doing a query mentioned above in a task. If I don't have it named "id", it works fine. Is something wrong with defining a field with "id = mongoengine.IntField(required=True)" ?

**Comment by lazarm** [comment]
Great, what a mess... just great! New pull request?

**Comment by mitar** [comment]
Yes, better a new one. Or clean this one. You can remove the commit you don't like from your repository and it will also remove here.

File: piplmesh/panels/bicikelj/models.py (hunk 9 above)
Comment (mitar) [inline piplmesh/panels/bicikelj/models.py]: Yes, `id` field is special. Maybe name it then `stand_id` (or `station_id`), I am not sure what you decided to use.
